src/main/java/org/traccar/sms/SnsSmsClient.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2021 Anton Tananaev (anton@traccar.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.sms;
+
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.services.sns.AmazonSNS;
+import com.amazonaws.services.sns.AmazonSNSClientBuilder;
+import com.amazonaws.services.sns.model.MessageAttributeValue;
+import com.amazonaws.services.sns.model.PublishRequest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.traccar.Context;
+import org.traccar.config.Keys;
+import org.traccar.notification.MessageException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SnsSmsClient implements SmsManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnsSmsClient.class);
+
+    private final AmazonSNS snsClient;
+
+    public SnsSmsClient() {
+        if (Context.getConfig().hasKey(Keys.SMS_AWS_REGION)
+                && Context.getConfig().hasKey(Keys.SMS_AWS_ACCESS)
+                && Context.getConfig().hasKey(Keys.SMS_AWS_SECRET)) {
+            snsClient = awsSNSClient();
+        } else {
+            throw new RuntimeException("SNS Not Configured Properly. Please provide valid config.");
+        }
+    }
+
+    public AmazonSNS awsSNSClient() {
+        BasicAWSCredentials awsCredentials = new BasicAWSCredentials(Context.getConfig().getString(Keys.SMS_AWS_ACCESS),
+                Context.getConfig().getString(Keys.SMS_AWS_SECRET));
+        return AmazonSNSClientBuilder.standard().withRegion(Context.getConfig().getString(Keys.SMS_AWS_REGION))
+                .withCredentials(new AWSStaticCredentialsProvider(awsCredentials)).build();
+    }
+
+    @Override
+    public void sendMessageSync(String destAddress, String message, boolean command)
+            throws InterruptedException, MessageException {
+        Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
+        smsAttributes.put("AWS.SNS.SMS.SenderID",
+                new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
+        smsAttributes.put("AWS.SNS.SMS.SMSType",
+                new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
+        snsClient.publish(new PublishRequest().withMessage(message)
```

Comment:
Wait. Is this method asynchronous or not?

Reply:
Yes, this method is `asynchronous`. While publishing the message AWS returns request id and behind the scenes use `asynchronous` processing for sending the actual message. (as far as i know, AWS is not providing any `synchronous` method for sending sns messages)

Reply:
OK, can you please provide some reference to confirm that it's `asynchronous`. Also, I'm not sure how you asynchronous method can return those exceptions.

Reply:
**Reference** : [AWS SNS Publish Message](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sns/AmazonSNSClient.html#publish-com.amazonaws.services.sns.model.PublishRequest-)

As described in the documentation, it just returns messageId and processes the request.

`When a messageId is returned, the message has been saved and Amazon SNS will attempt to deliver it shortly.`

Reply:
Just want to clarify. I'm talking about network request to Amazon SNS. I'm not talking about flow on the Amazon server. It sounds like the documentation you are referencing is referring to the latter.

Reply:
When it comes to network call, it is synchronous.

What we can do is, we can create another method and make it as Async and call `sendMessageSync(...)` method from the newly created method. - For this we will have to use `@Async` annotation which requires `spring-context` dependency.

Please share your thoughts on the same.

Reply:
I don't think we need to include Sprint to just make a single asynchronous method. I think just starting a new thread would be enough.

Reply:
I have updated PR with required changes.

Thanks,
Subodh